### PR TITLE
Fix gemma4 audio-encoder audio_features export rank (#570)

### LINKED ISSUE #570: [Model Request]: nvidia/magpie_tts_multilingual_357m
### Model name

nvidia/magpie_tts_multilingual_357m

### Hugging Face model or repository link

https://huggingface.co/nvidia/magpie_tts_multilingual_357m

### Primary task

Speech recognition or audio

### Use case

https://huggingface.co/blog/nvidia/magpie-tts-multilingual-voice-agents

### Execution providers

- [x] CPU
- [x] CUDA
- [x] QNN
- [x] OpenVINO
- [x] MIGraphX
- [x] MLX (macOS)
- [x] Other

### Target hardware

_No response_

### Quantization requirements

_No response_

### Contribution

- [ ] I can help test this model.
- [x] I can help implement this model.
- [x] I have checked that a duplicate request does not already exist.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by justinchuby**
## Problem (onnx-genai #570)
Every gemma4 (E2B/E4B) audio package is rejected at onnx-genai pipeline admission because the audio dataflow edge has mismatched ranks:

```
'audio_encoder.audio_features -> embedding.audio_features' has incompatible ranks:
producer rank 3, consumer rank 2.
```

The `audio_encoder` sub-model emitted rank-3 `[B, T//4, H]`, while `embedding` consumes rank-2 `[num_valid_frames, H]` (its scatter expects per-token rows). The shipped `gemma4-e2b-onnx` export contains **0 Compress ops** and ends at the projector MatMul — raw rank-3.

## HF reference semantics
HuggingFace gemma4 selects `audio_features[audio_mask]` — a boolean-mask gather that **both** drops padding frames **and** flattens the batch axis, producing rank-2 `[num_valid_frames, text_hidden_size]` rows that align 1:1 with audio placeholder tokens before they are scattered into the language-model embeddings.

## Fix
Move the mask-selection **into** `_Gemma4AudioEncoderModel.forward`: after the projector, flatten `[B, T//4, H] -> [-1, H]` and reuse the shared bf16-safe `_dtype_safe_compress` helper to emit rank-2 `[num_valid_frames, H]`, while still returning `downsampled_mask` for the `audio_features_mask` diagnostic output. The task builder `_build_audio` no longer duplicates the flatten+Compress.

- **DRY / general:** matches the encoder-free `_Gemma4UnifiedAudioEmbedderModel` (which already emitted rank-2) and covers **all** gemma4 Conformer-audio variants since they share `_Gemma4AudioEncoderModel` — no E2B-specific hack.
- The image edge is already rank-2 (`vision_encoder` flattens batch); no change there.
- Module docstrings updated to state the rank-2 contract.

## Test
New `TestGemma4AudioEncoderRank` (non-vacuous):
- asserts exported `audio_encoder.audio_features` rank == `embedding.audio_features` input rank == **2**, trailing dim == `hidden_size` on both ends;
- asserts a `Compress` node exists in the audio graph, so the reduction is a real mask-select, not a bare reshape that would misalign rows with placeholder tokens.

## Validation
- `pytest src/mobius/models/gemma4_test.py` → 20 passed (incl. 2 new)
- `pytest src/mobius/components/_gemma4_audio_test.py` → 31 passed
- `ruff check` on changed files → clean

Closes #570.

> Do not merge — mobius PRs are merged by Justin.

### CODE REVIEW

**Comment by justinchuby** [comment]
✅ Coordinator-verified, ready for your merge (not self-merged, per mobius policy).

Checked:
- **Root cause fix is correct:** the rank-3→2 mask-selection is moved *into* `_Gemma4AudioEncoderModel.forward` (flatten `[B,T//4,H]→[-1,H]` + `_dtype_safe_compress` by `downsampled_mask`), so the exported `audio_features` producer rank now matches the `embedding` sub-model's rank-2 consumer — closing the onnx-genai #570 admission rejection.
- **Matches HF semantics:** equivalent to `audio_features[audio_mask]` (rank-2 gather of valid frames), consistent with the documented gemma4 audio contract.
- **bf16-safe preserved:** the old inline `Cast(FLOAT)→Compress→CastLike` is replaced by the shared `_dtype_safe_compress` helper (same Cast-around-Compress), and the task-level duplicate is removed — DRY, general across E2B/E4B/-wp-a.
- **`downsampled_mask is None` branch** returns rank-3 only for eager unit calls; the export path raises if the mask is absent (`_gemma4.py:666`), so the consumer always sees rank-2.
- **Test is non-vacuous:** `TestGemma4AudioEncoderRank` asserts producer rank == consumer rank == 2 (trailing dim == hidden_size) *and* that a `Compress` node exists (real mask-select, not a bare reshape). ruff clean; gemma4 (20) + audio (31) tests pass.

Suggest a quick manual e2e sanity (export `gemma4-e2b` → confirm it now passes onnx-genai `pipeline_admission`) before merge, since we couldn't run the full cross-repo admission here.

**Comment by justinchuby** [comment]
This may break genai. Need to be careful

File: src/mobius/models/gemma4.py [hunk 1]
```diff
@@ -2639,9 +2639,19 @@ class _Gemma4AudioEncoderModel(nn.Module):
...
     Inputs:
     - ``input_features [B, T, input_size]``: mel-spectrogram features
+    - ``input_features_mask [B, T]``: BOOL mask, ``True`` for valid frames
 
     Output:
-    - ``audio_features [B, T//4, text_hidden_size]``: projected audio tokens
+    - ``audio_features [num_valid_frames, text_hidden_size]``: projected audio
+      tokens with padding frames stripped so output rows align 1:1 with audio
+      placeholder tokens.  This mirrors HF, which selects
+      ``audio_features[audio_mask]`` (a rank-2 gather) before scattering the
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/models/gemma4.py]: Fixed the docstring (forward is correct, so no code change). The Output section now describes the conditional rank:
- **with `input_features_mask`** (the export path — always supplied): rank-2 `[num_valid_frames, text_hidden_size]` after stripping padding frames;
- **without a mask** (`downsampled_mask is None`, e.g. a bare eager/unit call): the unmodified rank-3 `[B, T//4, text_hidden_size]` encoder output.

So call sites no longer assume rank-2 unconditionally.
